### PR TITLE
Bid schema, validation, & current price calculation

### PR DESCRIPTION
**Pull request by rstein66**
- Seller should not be able to place a bid on their own item  
- Buyer can place a bid IFF (1) the amount is higher than the current price and (2) the auction is open
- Buyer can place back-to-back bids

> Apologies for not noting & cleaning up the messy history prior to creating a PR!

### CODE REVIEW

**Comment by rstein66** [comment]
[Bid schema, validation, and current price calculation](https://trello.com/c/mpeuSx33/15-bid-schema-validation-and-current-price-calculation)

**Comment by rstein66** [comment]
[Setup Continuous Integration (CI) to run tests.](https://trello.com/c/oEIjKPC3/12-setup-continuous-integration-ci-to-run-tests)

File: lib/renaissance/auctions/auction.ex [hunk 1]
```diff
@@ -21,24 +23,15 @@ defmodule Renaissance.Auction do
     auction
     |> cast(attrs, @required_fields ++ @optional_fields)
     |> validate_required(@required_fields)
-    |> validate_price()
-    |> validate_date()
+    |> foreign_key_constraint(:seller_id, name: :auctions_user_id_fkey)
+    |> Constraint.amount_constraint(:price)
+    |> end_datetime_constraint()
```
Comment (kevbuchanan) [inline lib/renaissance/auctions/auction.ex]: `_constraint` implies something being applied in the database that Ecto should know about. This should be `validate_end_datetime`

File: lib/renaissance/bids/bids.ex [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    Bid.changeset(%Bid{}, params)
+    |> Repo.insert()
+  end
+
+  def exists?(id) do
+    Repo.exists?(from b in Bid, where: b.id == ^id)
+  end
+
+  def get_highest_bid(auction_id) do
+    query = "SELECT bidder_id, amount, created_at
```
Comment (kevbuchanan) [inline lib/renaissance/bids/bids.ex]: This query should be constructed through Ecto: https://hexdocs.pm/ecto/Ecto.Query.html

File: lib/renaissance/helpers/adapt.ex [hunk 3]
```diff
@@ -0,0 +1,19 @@
+defmodule Renaissance.Helpers.Adapt do
```
Comment (kevbuchanan) [inline lib/renaissance/helpers/adapt.ex]: The `Adapt` name is a little bit unclear. Maybe just `Renaissance.Format.Money`? Everything in here has to do with Money.

File: .gitignore [hunk 4]
```diff
@@ -245,3 +245,4 @@ bh_unicode_properties.cache
 GitHub.sublime-settings
 
 # End of https://www.gitignore.io/api/sublimetext
+assets/package-lock.json
```
Comment (ssunday) [inline .gitignore:248]: Why remove package-lock.json from VCS?

File: lib/renaissance/auctions/auction.ex (hunk 1 above)
Comment (ssunday) [reply lib/renaissance/auctions/auction.ex]: +1

File: lib/renaissance/helpers/compare.ex [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+      amount?(m2) == true && is_nil(m1) ->
+        return_money_max(m2)
+
+      true ->
+        nil
+    end
+  end
+
+  defp return_money_max(m) do
```
Comment (kevbuchanan) [inline lib/renaissance/helpers/compare.ex]: `to_money` might be a better name.

File: lib/renaissance/auctions/auctions.ex [hunk 6]
```diff
@@ -1,50 +1,51 @@
...
+  end
+
+  def get!(id) do
     Auction
     |> preload(:seller)
     |> Repo.get!(id)
   end
 
+  def get_seller_id(id) do
+    if exists?(id), do: get!(id).seller_id
```
Comment (ssunday) [inline lib/renaissance/auctions/auctions.ex]: I'm not 100% on the ecto optimization of it, but I'm guessing ecto pulls out the full record with `get!` in which case why have methods in the repo to pull specific things if it's not optimizing the query? Why not have a find_if_exists type call that does get and then in the methods get the attributes?

File: lib/renaissance/auctions/auctions.ex [hunk 7]
```diff
@@ -1,50 +1,51 @@
...
+  end
+
+  def get_starting_amount(id) do
+    if exists?(id), do: get!(id).price
+  end
+
+  def get_current_amount(id) do
+    larger =
+      Bids.get_highest_bid_amount(id)
+      |> Compare.money_max(get_starting_amount(id))
```
Comment (ssunday) [inline lib/renaissance/auctions/auctions.ex]: would a case piped into this be cleaner?

File: lib/renaissance/bids/bids.ex (hunk 2 above)
Comment (ssunday) [inline lib/renaissance/bids/bids.ex]: Does this have to be in raw SQL?

File: lib/renaissance/bids/bids.ex [hunk 8]
```diff
@@ -0,0 +1,53 @@
+defmodule Renaissance.Bids do
+  import Ecto.Query
+  alias Renaissance.{Repo, Bid}
+  alias Renaissance.Helpers.Adapt
+
+  def place_bid(params) do
+    params = Adapt.format_amount(params, "amount")
```
Comment (ssunday) [inline lib/renaissance/bids/bids.ex]: so this is basically creating it...do we need to have it use a special name instead of the standard ones?

File: lib/renaissance/helpers/compare.ex [hunk 9]
```diff
@@ -0,0 +1,33 @@
+defmodule Renaissance.Helpers.Compare do
+  alias Renaissance.Helpers.Adapt
+
+  def amount?(%Money{}), do: true
+
+  def amount?(p) do
```
Comment (ssunday) [inline lib/renaissance/helpers/compare.ex]: can't you just return `is_integer(p)`?

File: lib/renaissance/helpers/compare.ex [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    if is_integer(p), do: true, else: false
+  end
+
+  def money_max(m1, m2) do
+    cond do
+      amount?(m1) && amount?(m2) == true ->
+        if Adapt.money_value(m1) >= Adapt.money_value(m2) do
```
Comment (ssunday) [inline lib/renaissance/helpers/compare.ex]: could this be extracted out, it's hard to read

File: lib/renaissance/helpers/constraint.ex [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    else
+      message = ~s(can't bid on auction item that you're selling)
+      add_error(changeset, field, message)
+    end
+  end
+
+  def open_constraint(changeset, field) do
+    auction_id = get_change(changeset, :auction_id, 0)
+
+    if Auctions.open?(auction_id) in [nil, true] do
```
Comment (ssunday) [inline lib/renaissance/helpers/constraint.ex]: If the record doesn't exist....shouldn't that be a diff error?

File: priv/repo/migrations/20190501154649_transition_table_bids_timestamps.exs [hunk 12]
```diff
@@ -0,0 +1,19 @@
+defmodule Renaissance.Repo.Migrations.TransitionTableBidsTimestamps do
+  use Ecto.Migration
+
+  def up do
+    rename table(:bids), :inserted_at, to: :created_at
```
Comment (kevbuchanan) [inline priv/repo/migrations/20190501154649_transition_table_bids_timestamps.exs:5]: I'm curious why it was necessary to change the conventional `inserted_at` and `updated_at` for bids...?

File: test/renaissance/users/users_test.exs [hunk 13]
```diff
@@ -1,40 +1,49 @@
...
 
-  describe "users" do
+  describe "register_user/2" do
     test "stores valid" do
-      {:ok, user} = Users.register_user(@valid_attrs)
+      {:ok, user} = Users.register_user(@attrs)
 
-      assert user.email == @valid_attrs.email
-      assert user.password_hash != @valid_attrs.password
+      assert Repo.exists?(User) == true
```
Comment (kevbuchanan) [inline test/renaissance/users/users_test.exs]: `Repo.exists?(User)` will test that there is any user in the users table. I'm not sure if that's what's intended here. If so, a more clear and specific assertion would be nice.

File: test/renaissance/auctions/auctions_test.exs [hunk 14]
```diff
@@ -17,66 +17,79 @@ defmodule Renaissance.Test.AuctionsTest do
...
+      {:ok, auction} = Auctions.create_auction(seller_id, @auction_one)
+      assert Auctions.exists?(auction.id) == true
+    end
 
-      exception =
-        assert_raise Ecto.ConstraintError, fn ->
-          Auctions.create_auction(invalid_seller_id, @auction_two)
-        end
+    test "false when no auction with given id" do
+      assert Auctions.exists?(0) == false
```
Comment (ssunday) [inline test/renaissance/auctions/auctions_test.exs]: `asset_false` works, no?

File: lib/renaissance/auctions/auctions.ex (hunk 6 above)
Comment (kevbuchanan) [inline lib/renaissance/auctions/auctions.ex]: There seems to be this pattern of `exists?(thing), do: the_operation(thing)`, which isn't really necessary, since you can just handle nil from get:

```
case get(id) do
  nil -> nil
  auction -> auction.seller_id
end
```

File: priv/repo/migrations/20190501154649_transition_table_bids_timestamps.exs (hunk 12 above)
Comment (ssunday) [reply priv/repo/migrations/20190501154649_transition_table_bids_timestamps.exs:5]: ^^

File: lib/renaissance/auctions/auctions.ex (hunk 6 above)
Comment (ssunday) [reply lib/renaissance/auctions/auctions.ex]: IMO, should this even be checking it? Shouldn't checking for record existing be higher up the handling order?

File: lib/renaissance/helpers/compare.ex (hunk 10 above)
Comment (kevbuchanan) [inline lib/renaissance/helpers/compare.ex]: Can you use the `Money` comparison functions here? https://hexdocs.pm/money/Money.html#compare/2

File: lib/renaissance/helpers/constraint.ex [hunk 15]
```diff
@@ -0,0 +1,42 @@
+defmodule Renaissance.Helpers.Constraint do
+  import Ecto.Changeset
+  alias Renaissance.Helpers.Adapt
+  alias Renaissance.Auctions
+
+  def amount_constraint(changeset, field) do
+    auction_id = get_change(changeset, :auction_id, 0)
```
Comment (kevbuchanan) [inline lib/renaissance/helpers/constraint.ex]: How does this work from Auction, since Auction doesn't have an auction_id?

File: lib/renaissance/helpers/constraint.ex [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+    proposed = get_change(changeset, field, 0) |> Adapt.money_value()
+    current = Auctions.get_current_amount(auction_id) |> Adapt.money_value()
+
+    if proposed > current do
```
Comment (kevbuchanan) [inline lib/renaissance/helpers/constraint.ex]: This check is not guaranteed to be accurate with concurrent bids being placed. Is that an issue? And if so, how could you ensure that a valid bid must always be placed higher than the current highest bid?

File: lib/renaissance/helpers/constraint.ex [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+      changeset
+    else
+      message = ~s(must be greater than #{Money.new(current)})
+      add_error(changeset, field, message)
+    end
+  end
+
+  def bidder_constraint(changeset, field) do
+    bidder_id = get_change(changeset, field)
+    auction_id = get_change(changeset, :auction_id, 0)
```
Comment (kevbuchanan) [inline lib/renaissance/helpers/constraint.ex]: Why do we default to 0 here?

File: lib/renaissance/helpers/constraint.ex [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+    seller_id = Auctions.get_seller_id(auction_id)
+
+    if is_nil(seller_id) or bidder_id != seller_id do
```
Comment (kevbuchanan) [inline lib/renaissance/helpers/constraint.ex]: This condition is a little bit hard to parse. I guess I would expect a little bit more clarity on the condition that we actually care about here:

```
case {seller_id, bidder_id} do
  {nil, _} -> changeset # How do we even end up in this place?
  {seller_id, bidder_id} when seller_id == bidder_id -> add_error(changeset, "can't bid on the item you're selling")
  _ -> changeset
end
```

File: lib/renaissance/helpers/constraint.ex (hunk 11 above)
Comment (kevbuchanan) [inline lib/renaissance/helpers/constraint.ex]: The fact that `open?` returns nil is a bit weird. `something?` should only return true or false.

Comment (ssunday) [reply lib/renaissance/helpers/constraint.ex]: agreed

File: .gitignore (hunk 4 above)
Comment (xray) [reply .gitignore:248]: I found it to be adding a lot of noise to diffs, but I'll just commit it separately instead.

File: priv/repo/migrations/20190501154649_transition_table_bids_timestamps.exs (hunk 12 above)
Comment (rstein66) [reply priv/repo/migrations/20190501154649_transition_table_bids_timestamps.exs:5]: I will remove this. 

_[I had changed it because, in my mind, having only one timestamp–`created_at`–makes it clearer that the entries in `bids` will not be modified, which contrasts with the entries in `users` & `auctions`.]_

File: test/renaissance/auctions/auctions_test.exs (hunk 14 above)
Comment (rstein66) [reply test/renaissance/auctions/auctions_test.exs]: Resolved with 46d8ca2 & 209badd.

_(Personally, I find [`refute`](https://hexdocs.pm/ex_unit/ExUnit.Assertions.html#functions) less clear, especially because it checks if the value is `in [nil, false]`, but that's just me.)_

File: lib/renaissance/bids/bids.ex (hunk 8 above)
Comment (rstein66) [reply lib/renaissance/bids/bids.ex]: Resolved in 15d9389.

File: priv/repo/migrations/20190501154649_transition_table_bids_timestamps.exs (hunk 12 above)
Comment (kevbuchanan) [reply priv/repo/migrations/20190501154649_transition_table_bids_timestamps.exs:5]: I think that's fine and makes sense.

File: lib/renaissance/bids/bids.ex (hunk 2 above)
Comment (rstein66) [reply lib/renaissance/bids/bids.ex]: Resolved in 845f45e.

Comment (rstein66) [reply lib/renaissance/bids/bids.ex]: Resolved in 845f45e.

File: lib/renaissance/helpers/constraint.ex (hunk 16 above)
Comment (rstein66) [reply lib/renaissance/helpers/constraint.ex]: I don't think it's an issue, at least based on my current understanding, which is as follows:
- If there are concurrent [valid] bids of difference amounts, then (for all practical purposes) it's as if the users' with the lower bids were out-bid, and therefore no special handling is needed. 
-  I updated `get_highest_bid` ([updated query](https://github.com/xray/Renaissance/blob/845f45ee2e/lib/renaissance/bids/bids.ex#L30);  [test](https://github.com/xray/Renaissance/blob/845f45ee2e/test/renaissance/bids/bids_test.exs#L117)) to return the earliest (or at least first recorded) bid if there are multiple valid bids with the highest amount for a given auction so that, if the auction closes in this state, the winner will be the first person to place the bid.

File: lib/renaissance/auctions/auctions.ex (hunk 6 above)
Comment (rstein66) [reply lib/renaissance/auctions/auctions.ex]: I wouldn't be at all surprised if that's true & can see the value, but I don't know if it would be worthwhile at present. As I see it, at present, this is the only function that retrieved a record, sans preloading, & extracts a specific field without adding additional functionality.

Comment (rstein66) [reply lib/renaissance/auctions/auctions.ex]: Where do you mean by higher up? The form?

If yes, the form errors (beyond those that can be checked using HTML tags) are generated by the changeset validation. I'm unaware of how else to generate them. All the examples I've seen generate form errors by means of the changeset validation; that said, I have not researched this topic in any depth.

**Comment by xray** [comment]
[User can bid on an item. (8.25pts)](https://trello.com/c/3N9qLZTl/9-user-can-bid-on-an-item-825pts)

**Comment by xray** [comment]
[User can bid on an item. (8.25pts)](https://trello.com/c/3N9qLZTl/9-user-can-bid-on-an-item-825pts)
